classes/observers/traits/UsageEvent.inc.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/**
+ * @file classes/observers/events/UsageEvent.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class UsageEvent
+ * @ingroup observers_traits
+ *
+ * @brief Usage event trait.
+ *
+ */
+
+namespace PKP\observers\traits;
+
+use APP\core\Application;
+use APP\core\Request;
+use APP\submission\Submission;
+use Illuminate\Foundation\Events\Dispatchable;
+use PKP\config\Config;
+use PKP\context\Context;
+use PKP\core\Core;
+use PKP\core\Registry;
+use PKP\submission\Representation;
+use PKP\submissionFile\SubmissionFile;
+
+trait UsageEvent
+{
+    use Dispatchable;
+
+    /** Current time */
+    public string $time;
+
+    /** Viewed/downloaded object, one of the Application::ASSOC_TYPE_... constants */
+    public int $assocType;
+
+    /** Canonical URL for the pub object */
+    public string $canonicalUrl;
+
+    public Request $request;
+
+    public Context $context;
+
+    public ?Submission $submission;
+
+    /** Representation (galley or publication format) */
+    public ?Representation $representation;
+
+    public ?SubmissionFile $submissionFile;
+
+    /** Application's complete version string */
+    public string $version;
+
+    /**
+     * Create a new usage event instance.
+     */
+    protected function constructUsageEvent(int $assocType, Context $context, Submission $submission = null, Representation $representation = null, SubmissionFile $submissionFile = null)
```

Comment:
These are all type-hinted and can't be swapped accidentally without causing an error, so no need to change it, but this is a case where you could also consider using named arguments.